Kill validateDeclForNameLookup Harder

CodaFi opened the pull request:
Using the techniques from @xymus' patch, try to kill validateDeclForNameLookup harder.  There are some changes from the exact code in #24957 that, unfortunately, I cannot get to pass the tests to land independently.  Namely, this patch

- Defines a new UnderlyingTypeRequest for computing the interface type of the underlying type if we have it.
- Changes the structural type request to resugar with a `TypeAliasType`.  I suspect this is what is causing the changes in requirement signatures in the few tests that are touched here.
- Separate computing the interface type of a TypeAliasDecl from computing its underlying type.  Not doing this makes reasoning about the laziness in UnderlyingTypeRequest interacting with the interface type of the typealias extremely difficult.

All of this should finally let us remove `validateDeclForNameLookup`.  I have great fears that this will fail the compatibility suite.  But there's only one way to find out!

CodaFi committed:
Define UnderlyingTypeRequest

Define a request for computing the interface type of the underlying type of a typealias.  This can be used in place of the declared interface type of the alias itself when it is nested to preserve the fragile validation ordering issues that can cause.

CodaFi committed:
Resugar the structural type

Structural type computation ought to actually return a typealias type here using the same mechanism we compute the interface type with (minus the metatype).

CodaFi committed:
Improve the TypeAlias{Type|Decl} dump

Dump the underlying type by desugaring ourselves a little.

CodaFi committed:
Separate computing interface types and underlying types

Computing the interface type of a typealias used to push validation forward and recompute the interface type on the fly.  This was fragile and inconsistent with the way interface types are computed in the rest of the decls.  Separate these two notions, and plumb through explicit interface type computations with the same "computeType" idiom.  This will better allow us to identify the places where we have to force an interface type computation.

Also remove access to the underlying type loc.  It's now just a cache location the underlying type request will use.  Push a type repr accessor to the places that need it, and push the underlying type accessor for everywhere else.  Getting the structural type is still preferred for pre-validated computations.

This required the resetting of a number of places where we were - in many cases tacitly - asking the question "does the interface type exist".  This enables the removal of validateDeclForNameLookup

CodaFi committed:
Remove validateDeclForNameLookup

Flush it and the early validation hack now that we can delay computing the underlying interface type on demand and have taught type resolution to honor the structural type of a typealias.

This changes the way requirement signatures are spelled as a side effect.

CodaFi commented:
The canary in the coal mine:

@swift-ci please smoke test

CodaFi commented:
I have high hopes for this patch.  If it actually works then we can merge the lazy generic signatures patch minus a ton of the hacks.

CodaFi commented:
Oh joy, I need LLDB patches.

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -910,15 +910,44 @@ RequirementRequest::evaluate(Evaluator &evaluator,
+  // to compute the structural type.  This also prevents us from writing
+  // ErrorTypes into otherwise valid ASTs with generated typealiases.
+  if (typeAlias->hasInterfaceType()) {
+    return typeAlias->getInterfaceType()->getMetatypeInstanceType();
```
This should use getDeclaredInterfaceType()

CodaFi replied:
Good catch

slavapestov commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -3785,9 +3785,8 @@ PotentialArchetype *GenericSignatureBuilder::realizePotentialArchetype(
-
+    if (typealias->hasInterfaceType())
+      return typealias->getDeclaredInterfaceType();
     return typealias->getStructuralType();
```
Doesn't getStructuralType() subsume the above 'if' now?

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -1791,9 +1791,10 @@ resolveTypeDeclsToNominal(Evaluator &evaluator,
-        if (auto type = typealias->getUnderlyingTypeLoc().getType()) {
-          if (type->isAnyObject())
-            anyObject = true;
+        if (typealias->hasInterfaceType()) {
```
The hasInterfaceType() check should not be necessary?

CodaFi replied:
Requesting the underlying type potentially calls back into validation.  This is also justified because the interface type separation changes mean that `AnyObject` declarations will always have their interface type set by the time we see them.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4510,11 +4516,20 @@ ExtendedTypeRequest::evaluate(Evaluator &eval, ExtensionDecl *ext) const {
-               : extendedNominal->getDeclaredType();
+      // Nested Hack to break cycles if this is called before validation has
+      // finished.
+      if (aliasDecl->hasInterfaceType()) {
```
Instead of checking for hasInterfaceType(), validate the alias to make this deterministic

CodaFi replied:
This is here because we're validating the alias when this is false...

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -528,26 +525,39 @@ Type TypeChecker::resolveTypeInContext(
+                    dyn_cast<TypeAliasDecl>(unboundGeneric->getAnyGeneric())) {
+              if (ugAliasDecl == aliasDecl) {
+                if (resolution.getStage() == TypeResolutionStage::Structural &&
+                    !aliasDecl->hasInterfaceType()) {
```
Shouldn't need this check

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -528,26 +525,39 @@ Type TypeChecker::resolveTypeInContext(
+                  dyn_cast<TypeAliasType>(extendedType.getPointer())) {
+            if (aliasType->getDecl() == aliasDecl) {
+              if (resolution.getStage() == TypeResolutionStage::Structural &&
+                  !aliasDecl->hasInterfaceType()) {
```
Ditto

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -566,9 +576,13 @@ Type TypeChecker::resolveTypeInContext(
-      // Otherwise, simply return the underlying type.
+      // Otherwise, return the appropriate type.
+      if (resolution.getStage() == TypeResolutionStage::Structural &&
+          !aliasDecl->hasInterfaceType()) {
```
Ditto

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -595,13 +609,12 @@ Type TypeChecker::resolveTypeInContext(
-    selfType = resolution.mapTypeIntoContext(
-      foundDC->getDeclaredInterfaceType());
+    selfType =
+        resolution.mapTypeIntoContext(foundDC->getDeclaredInterfaceType());
```
There's a lot of re-formatting here which makes the patch harder to review. Is it really necessary?

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -973,28 +985,36 @@ static Type resolveTypeDecl(TypeDecl *typeDecl, SourceLoc loc,
+      !prevalidatingAlias(typeDecl, resolution)) {
     // Validate the declaration.
     if (lazyResolver)
       lazyResolver->resolveDeclSignature(typeDecl);
```
You could try removing this altogether, and validating the decl only once you know you need the interface type.

CodaFi replied:
There's some deep cycles at work that break the standard library if this is removed.

slavapestov commented:
Once getGenericSignature() is a request, we can request-ify getDeclaredInterfaceType() on a type alias decl.

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please smoke test

CodaFi commented:
Looks like changing the requirement signature has changed the way the exportability checker behaves.  Looking into this.

CodaFi committed:
Setting the underlying type should set the structural type too

This avoids making the structural type dependent on whether the interface type has been computed and neatly avoids special-casing non-parsed declarations which set their underlying type up front.

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please smoke test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please smoke test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please smoke test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please smoke test

slavapestov commented:
@CodaFi I see there's a few diagnostic checks in validateType(). They should be pushed down to resolveType() so that we emit the same diagnostics everywhere; validateType() should just be a transitional utility method that writes the Type back into a TypeLoc. But you don't have to fix that now.

Let's look at those cycles involving validateDecl() next -- we'll have to resolve them once getInterfaceType() becomes a request, because now cycles will produce diagnostics instead of silently bailing out of validation.

CodaFi commented:
@swift-ci please test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test source compatibility

CodaFi commented:
Sent the wrong CI command here.  Gonna try again when macOS fails.

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test source Linux platform

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test Linux platform

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test macOS platform

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test source compatibility

CodaFi committed:
Break some cycles

Use the interface type of the underlying type in the GSB in as many
cases as possible except for one important one: requirement signature
computation.  There, use the structural type so we don't wind up
recursively validating the requirements of an incomplete protocol.

CodaFi commented:
Canary

@swift-ci please smoke test

CodaFi commented:
Canary

apple/swift-lldb#1986
@swift-ci Please smoke test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test source compatibility

CodaFi commented:
Ugh, the same set of offenders as last time...

CodaFi committed:
Change a cycle condition

Not having the generic signature is the real culprit here.

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test source compatibility

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test Linux platform

CodaFi commented:
apple/swift-lldb#1986
@swift-ci Please test macOS platform

CodaFi commented:
⛵️

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3812,9 +3814,13 @@ void TypeChecker::validateDecl(ValueDecl *D) {
-    validateTypealiasType(*this, typeAlias);
+
+    // Finally, set the interface type.
+    if (!typeAlias->hasInterfaceType())
```
Is it ever the case that hasInterfaceType() returns true here? If you copy and pasted this from the old code it's just an artifact of validateDecl() being called after validateDeclForNameLookup()

CodaFi replied:
It was true when the code was written - something something cycles with witness matching.

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -556,7 +556,7 @@ static Type formExtensionInterfaceType(
 
   // If we have a typealias, try to form type sugar.
-  if (typealias && TypeChecker::isPassThroughTypealias(typealias)) {
+  if (typealias && TypeChecker::isPassThroughTypealias(typealias, typealias->getUnderlyingType(), nominal)) {
```
This line is too long

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -2287,10 +2287,22 @@ class swift::DeclDeserializer {
 
+    auto underlying = MF.getType(underlyingTypeID);
+    alias->setUnderlyingType(underlying);
+    SubstitutionMap subs;
```
Is this not `alias->computeType()`?

slavapestov reviewed (commented):
Nice work @xymus and @CodaFi! Really happy to see this land.

CodaFi commented on lib/AST/ASTDumper.cpp:
```diff
@@ -616,9 +616,9 @@ namespace {
       printCommon(TAD, "typealias");
       PrintWithColorRAII(OS, TypeColor) << " type='";
-      if (TAD->getUnderlyingTypeLoc().getType()) {
+      if (auto underlying = TAD->getUnderlyingType()) {
```
I'm going to add an RAII type to catch all of these.
